tests/unittests/test_updated_bookmark.py

```diff
@@ -37,3 +37,53 @@ def test_update_bookmark(self, mock_now, mock_get_query_params, mock_update_book
         customers = list(CUSTOMER_OBJECT.get_objects())
         # Verify that the min-max evaluation returns correct results and provided in the update_bookmark()
         mock_update_bookmark.assert_called_with(strftime(strptime_to_utc(NOW_TIME) - datetime.timedelta(DATE_WINDOW_SIZE)))
+
+
+    @mock.patch("tap_shopify.streams.base.Stream.call_api")
+    @mock.patch("tap_shopify.streams.base.Stream.get_bookmark", return_value=strptime_to_utc('2021-08-11T01:55:05-04:00'))
+    @mock.patch("tap_shopify.streams.base.Stream.update_bookmark")
+    @mock.patch("tap_shopify.streams.base.Stream.get_query_params")
+    @mock.patch("singer.utils.now", return_value=strptime_to_utc(NOW_TIME))
+    def test_since_id_and_updated_at_max_deleted(self, mock_now, mock_get_query_params, mock_update_bookmark, mock_get_bookmark, mock_call_api):
+        """Verify after successful sync since_id and updated_at_max keys are deleted from the state"""
+        mock_call_api.return_value = [CUSTOMER_1, CUSTOMER_2]
+        Context.state = {"bookmarks": {
+            "currently_sync_stream": 'customers',
+            "customers": {
+                "updated_at": "2021-03-27T00:00:00.000000Z",
+                "since_id": 15,
+                "updated_at_max": "2021-04-26T00:00:00.000000Z"}}}
+
+        customers = list(CUSTOMER_OBJECT.get_objects())
+
+        self.assertNotIn("since_id", Context.state["bookmarks"]["customers"])
+        self.assertNotIn("updated_at_max", Context.state["bookmarks"]["customers"])
+
+    @mock.patch("tap_shopify.streams.base.Stream.call_api")
+    @mock.patch("tap_shopify.streams.base.Stream.get_bookmark", return_value=strptime_to_utc('2021-08-11T01:55:05-04:00'))
+    @mock.patch("tap_shopify.streams.base.Stream.update_bookmark")
+    @mock.patch("tap_shopify.streams.base.Stream.get_query_params")
+    @mock.patch("singer.utils.now", return_value=strptime_to_utc(NOW_TIME))
+    def test_interrupted_sync(self, mock_now, mock_get_query_params, mock_update_bookmark, mock_get_bookmark, mock_call_api):
+        """Verify if sync is interrrupted twice in a row then since_id and updated_at_max keys are not deleted from the state"""
```

Comment:
This test seems poorly implemented.
1. We set the state
2. We mock all of the code that would alter the state
3. We assert the state is what we expect it to be

Shouldn't we start with an empty state, mock the API call so that state looks like
```python
{"bookmarks": {
            "currently_sync_stream": 'customers',
            "customers": {
                "updated_at": "2021-03-27T00:00:00.000000Z",
                "since_id": 15,
                "updated_at_max": "2021-04-26T00:00:00.000000Z"}}}
```
then force the tap to throw an exception, and then assert the state has a `since_id` and an `updated_at_max`?

Reply:
In the code implementation, we are poping `since_id` and `updated_at_max`. This test is verifying that we are not poping these intermediate bookmarks too early before we successfully sync till `updated_at_max` in the bookmark. Early deletion of intermitent bookmarks will cause the undesired results in case sync is interrupted.